lib/s3routes/routesUtils.ts

```diff
@@ -441,23 +447,42 @@ function retrieveData(
                 });
                 // errors on server side with readable stream
                 readable.on('error', err => {
-                    log.error('error piping data from source');
+                    log.error('error piping data from source', { error: err });
                     _destroyResponse();
                     return cbOnce(err);
                 });
+                // At this point, we assume that the data backend is available.
+                // We write the http headers before piping the data to the
+                // response object.
+                // Any subsequent error leads to destror deletion.
+                // todo: need to check if headers were already set
+                if (!response.headersSent) {
+                    response.writeHead(httpCodeOnSucess);
+                }
+                // TODO ARSN-474: can this be moved above, so we
+                // do not need to call _destroyReadable in the error case
+                // in the callback of data.get()? The 'close' event is not
+                // called if 'end' is.
                 currentStream = readable;
                 return readable.pipe(response, { end: false });
             }), err => {
-            currentStream = null;
-            if (err) {
-                log.debug('abort response due to error', {
-                    // @ts-expect-error
-                    error: err.code, errMsg: err.message });
-            }
-            // call end for all cases (error/success) per node.js docs
-            // recommendation
-            response.end();
-        },
+                currentStream = null;
+                if (err) {
+                    log.error('abort response due to error', {
+                        // @ts-expect-error
+                        error: err.code, errMsg: err.message });
+                    if (!response.headersSent) {
+                        return XMLResponseBackend.errorResponse(errors.ServiceUnavailable, response, log);
+                    }
+                }
+                // call end for all cases (error/success) per node.js docs
+                // recommendation
+                response.end(() => {
+                    log.end().info('responded with streamed content', {
```

Comment:
when this happens after an error, the httpCode is not relevant anymore : the log does not match in that case...
It is also somewhat redundant with the error messages posted above.

Reply:
See first comment's answser